crates/sui-types/src/multisig.rs

```diff
@@ -0,0 +1,316 @@
+    #[serde(skip)]
+    bytes: OnceCell<Vec<u8>>,
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl PartialEq for MultiSignature {
+    fn eq(&self, other: &Self) -> bool {
+        self.sigs == other.sigs && self.bitmap == other.bitmap && self.multi_pk == other.multi_pk
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Eq for MultiSignature {}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Hash for MultiSignature {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.as_ref().hash(state);
+    }
+}
+
+impl AuthenticatorTrait for MultiSignature {
+    fn verify_secure_generic<T>(
+        &self,
+        value: &IntentMessage<T>,
+        author: SuiAddress,
+    ) -> Result<(), SuiError>
+    where
+        T: Serialize,
+    {
+        if self.multi_pk.pk_map.len() > MAX_SIGNER_IN_MULTISIG {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid number of public keys".to_string(),
+            });
+        }
+
+        if <SuiAddress as From<MultiPublicKey>>::from(self.multi_pk.clone()) != author {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid address".to_string(),
+            });
+        }
+        let mut weight_sum = 0;
+        let message = bcs::to_bytes(&value).expect("Message serialization should not fail");
+
+        for (sig, i) in self.sigs.iter().zip(&self.bitmap) {
+            let pk_map =
+                self.multi_pk
+                    .pk_map
+                    .get(i as usize)
+                    .ok_or(SuiError::InvalidSignature {
+                        error: "Invalid public keys index".to_string(),
+                    })?;
+            let res = match sig {
+                CompressedSignature::Ed25519(s) => {
+                    let pk = Ed25519PublicKey::from_bytes(pk_map.0.as_ref()).map_err(|_| {
+                        SuiError::InvalidSignature {
+                            error: "Invalid public key".to_string(),
+                        }
+                    })?;
+                    pk.verify(&message, s)
+                }
+                CompressedSignature::Secp256k1(s) => {
+                    let pk = Secp256k1PublicKey::from_bytes(pk_map.0.as_ref()).map_err(|_| {
+                        SuiError::InvalidSignature {
+                            error: "Invalid public key".to_string(),
+                        }
+                    })?;
+                    pk.verify(&message, s)
+                }
+                CompressedSignature::Secp256r1(s) => {
+                    let pk = Secp256r1PublicKey::from_bytes(pk_map.0.as_ref()).map_err(|_| {
+                        SuiError::InvalidSignature {
+                            error: "Invalid public key".to_string(),
+                        }
+                    })?;
+                    pk.verify(&message, s)
+                }
+            };
+            if res.is_ok() {
```

Comment:
maybe we should fail when one of the sigs fails, even if the threshold of correct sigs is met; this is to avoid appending noise to the signature payload and spamming our network